File: scripts/normality_test.R
#!/usr/bin/env Rscript
# ============================================
# 正态性检验脚本
# medical-statistics skill
# ============================================

suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(haven))
suppressPackageStartupMessages(library(readxl))

args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  params <- list()
  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data" && i + 1 <= length(args)) {
      params$data <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--vars" && i + 1 <= length(args)) {
      params$vars <- unlist(strsplit(args[i + 1], ","))
      i <- i + 2
    } else {
      i <- i + 1
    }
  }
  return(params)
}

params <- parse_args(args)

if (is.null(params$data)) {
  stop("请指定数据文件路径: --data <path>")
}

# 读取数据
ext <- tolower(tools::file_ext(params$data))
if (ext == "csv") {
  df <- read.csv(params$data, stringsAsFactors = FALSE)
} else if (ext %in% c("xls", "xlsx")) {
  df <- read_excel(params$data)
} else if (ext %in% c("sav")) {
  df <- read_sav(params$data)
} else if (ext %in% c("dta")) {
  df <- read_dta(params$data)
} else {
  stop("不支持的文件格式: ", ext)
}

# 自动识别连续变量
if (is.null(params$vars)) {
  continuous_vars <- names(df)[sapply(df, function(x) is.numeric(x) && length(unique(x)) > 10)]
} else {
  continuous_vars <- params$vars[params$vars %in% names(df)]
}

cat("\n═══════════════════════════════════════════\n")
cat("        正态性检验结果\n")
cat("═══════════════════════════════════════════\n\n")

results <- data.frame(
  变量 = character(),
  样本量 = integer(),
  均值 = numeric(),
  标准差 = numeric(),
  中位数 = numeric(),
  Q1 = numeric(),
  Q3 = numeric(),
  偏度 = numeric(),
  峰度 = numeric(),
  Shapiro_P = numeric(),
  正态性判断 = character(),
  stringsAsFactors = FALSE
)

for (var in continuous_vars) {
  x <- df[[var]]
  x <- x[!is.na(x)]
  n <- length(x)

  if (n < 3) {
    cat(sprintf("  ⚠ %s: 样本量不足 (%d)，跳过\n", var, n))
    next
  }

  mean_val <- mean(x, na.rm = TRUE)
  sd_val <- sd(x, na.rm = TRUE)
  median_val <- median(x, na.rm = TRUE)
  q1 <- quantile(x, 0.25, na.rm = TRUE)
  q3 <- quantile(x, 0.75, na.rm = TRUE)

  # 偏度峰度
  skewness <- sum((x - mean_val)^3) / (n * sd_val^3)
  kurtosis <- sum((x - mean_val)^4) / (n * sd_val^4) - 3

  # Shapiro-Wilk 检验
  if (n < 5000 && n >= 3) {
    sw_test <- shapiro.test(x)
    shapiro_p <- sw_test$p.value
  } else if (n >= 5000) {
    # KS 检验
    ks_test <- ks.test(x, "pnorm", mean_val, sd_val)
    shapiro_p <- ks_test$p.value
  } else {
    shapiro_p <- NA
  }

  # 判断
  if (!is.na(shapiro_p)) {
    if (shapiro_p > 0.05) {
      normality <- "正态分布"
    } else if (shapiro_p > 0.01) {
      normality <- "弱非正态"
    } else {
      normality <- "非正态分布"
    }
  } else {
    normality <- "无法判断"
  }

  results <- rbind(results, data.frame(
    变量 = var,
    样本量 = n,
    均值 = round(mean_val, 3),
    标准差 = round(sd_val, 3),
    中位数 = round(median_val, 3),
    Q1 = round(q1, 3),
    Q3 = round(q3, 3),
    偏度 = round(skewness, 3),
    峰度 = round(kurtosis, 3),
    Shapiro_P = round(shapiro_p, 4),
    正态性判断 = normality,
    stringsAsFactors = FALSE
  ))

  # 输出单个变量详情
  cat(sprintf("  ■ %s (n=%d)\n", var, n))
  cat(sprintf("    正态性检验 P = %.4f\n", shapiro_p))
  if (shapiro_p > 0.05) {
    cat(sprintf("    ✅ 正态分布 (P > 0.05)\n"))
    cat(sprintf("    描述方式: Mean ± SD = %.2f ± %.2f\n", mean_val, sd_val))
  } else {
    cat(sprintf("    ❌ 非正态分布 (P ≤ 0.05)\n"))
    cat(sprintf("    描述方式: Median (Q1, Q3) = %.2f (%.2f, %.2f)\n", median_val, q1, q3))
  }
  cat("    -----------------------------------\n")
}

# 输出汇总表
cat("\n━━━━━━━━━━━━━━━━━━ 汇总 ━━━━━━━━━━━━━━━━━━\n")
print(results, row.names = FALSE)

# 输出建议的描述方式和检验方法
cat("\n━━━━━━━━━━━━━━━━━━ 统计方法建议 ━━━━━━━━━━━━━━━━━━\n")
for (i in seq_len(nrow(results))) {
  if (results$正态性判断[i] %in% c("正态分布", "弱非正态")) {
    desc <- sprintf("Mean ± SD (%.2f ± %.2f)", results$均值[i], results$标准差[i])
    test <- "t检验 / ANOVA (参数检验)"
  } else {
    desc <- sprintf("Median (Q1, Q3) = %.2f (%.2f, %.2f)", results$中位数[i], results$Q1[i], results$Q3[i])
    test <- "Wilcoxon秩和检验 / Kruskal-Wallis检验 (非参数检验)"
  }
  cat(sprintf("  %s:\n", results$变量[i]))
  cat(sprintf("    描述: %s\n", desc))
  cat(sprintf("    推荐方法: %s\n", test))
}

cat("\n═══════════════════════════════════════════\n")

File: scripts/normality_test.py
#!/usr/bin/env python3
"""
Normality Assessment Script (Python)
Medical Statistics Skill
"""
import argparse
import sys
import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt


def read_data(filepath):
    ext = filepath.split(".")[-1].lower()
    if ext == "csv":
        return pd.read_csv(filepath)
    elif ext in ("xls", "xlsx"):
        return pd.read_excel(filepath)
    elif ext == "sav":
        import pyreadstat
        return pd.read_spss(filepath)
    elif ext == "dta":
        return pd.read_stata(filepath)
    else:
        raise ValueError(f"Unsupported format: {ext}")


def normality_report(x, var_name):
    x = x.dropna()
    n = len(x)
    if n < 3:
        return None

    mean_val = x.mean()
    sd_val = x.std()
    median_val = x.median()
    q1, q3 = x.quantile(0.25), x.quantile(0.75)
    skew = stats.skew(x)
    kurt = stats.kurtosis(x, fisher=True)

    if n < 5000:
        w, p = stats.shapiro(x)
        test_name = "Shapiro-Wilk"
    else:
        from scipy.stats import kstest
        stat, p = kstest(x, "norm", args=(mean_val, sd_val))
        w = stat
        test_name = "Kolmogorov-Smirnov"

    if p > 0.05:
        normality = "Normal"
        desc = f"Mean ± SD = {mean_val:.2f} ± {sd_val:.2f}"
        rec_method = "Parametric (t-test / ANOVA)"
    else:
        normality = "Non-normal"
        desc = f"Median (Q1, Q3) = {median_val:.2f} ({q1:.2f}, {q3:.2f})"
        rec_method = "Non-parametric (Wilcoxon / Kruskal-Wallis)"

    return {
        "variable": var_name, "n": n,
        "mean": round(mean_val, 3), "sd": round(sd_val, 3),
        "median": round(median_val, 3), "Q1": round(q1, 3), "Q3": round(q3, 3),
        "skewness": round(skew, 3), "kurtosis": round(kurt, 3),
        "test": test_name, "statistic": round(w, 4), "P": round(p, 4),
        "normality": normality, "description": desc,
        "recommended_method": rec_method
    }


def main():
    parser = argparse.ArgumentParser(description="Normality Assessment")
    parser.add_argument("--data", required=True, help="Data file path")
    parser.add_argument("--vars", help="Comma-separated variable names")
    args = parser.parse_args()

    df = read_data(args.data)
    print("\n" + "=" * 55)
    print("         Normality Assessment Results")
    print("=" * 55 + "\n")

    if args.vars:
        continuous_vars = [v.strip() for v in args.vars.split(",") if v.strip() in df]
    else:
        continuous_vars = [c for c in df.select_dtypes(include=[np.number]).columns
                           if df[c].nunique() > 10]

    results = []
    for var in continuous_vars:
        res = normality_report(df[var], var)
        if res is None:
            continue
        results.append(res)

        # Detailed output
        print(f"  ■ {var} (n={res['n']})")
        print(f"    {res['test']} P = {res['P']:.4f}")
        if res['normality'] == "Normal":
            print(f"    ✅ Normal (P > 0.05)")
        else:
            print(f"    ❌ Non-normal (P ≤ 0.05)")
        print(f"    {res['description']}")
        print("    " + "-" * 35)

    # Summary table
    print("\n" + "─" * 55)
    print(" Summary")
    print("─" * 55)
    res_df = pd.DataFrame(results)
    print(res_df.to_string(index=False,
          columns=["variable", "n", "P", "normality", "recommended_method"]))

    # Method recommendations
    print("\n" + "─" * 55)
    print(" Method Recommendations")
    print("─" * 55)
    for r in results:
        print(f"  {r['variable']}: {r['recommended_method']}")

    print("\n" + "=" * 55)


if __name__ == "__main__":
    main()
